### ISSUE TITLE
SASmodels - better solution for namespaces around the individual c-codes

### ISSUE THREAD

**Issue by willend** [McStasMcXtrace/McCode#830]
In a recent discussion on mcstas-users (https://mailman2.mcstas.org/pipermail/mcstas-users/2019q3/001301.html) it looks relevant to be able to combine more than one sasmodel in the same instrument.

I propose that we make an alternative python-script based solution that generates one component from one model, directly embedding the (rewritten) model c-code with mode-specific names for these functions:

<pre>
grep ^float sas_star_polymer.c

float form_volume(void);   ->   form_volume_star_polymer(void); 
float Iq(float q, float radius2, float arms);   ->   Iq_star_polymer(float q, float radius2, float arms);
float Iqxy(float qx, float qy, float radius2, float arms); -> ...
float form_volume(void) -> ...
float Iq(float q, float radius2, float arms) -> ...
float Iqxy(float qx, float qy, float radius2, float arms) -> ...
</pre>

I.e. one "flat" component file generated pr. sasmodel c-snippet. Perhaps as a user-level script to generate only the ones you are yourself interested of.

@trnielsen, @ebknudsen, @climbcat what do you think?

**Comment by ebknudsen** [comment]
Humm - I haven't thought this through completely ( I _am_ on vacation) but what if we did something along the lines of the mag-fields. I.e. a set of files .c and .h (autogenerated preferrably) containg the model functions and used a flat single component that contains an array of static function pointers,
Given the model number we could then call the function in the array that corresponds to the SAS-models number with a void pointer to it's set of parameters. That would avoid having lots of similar comps while still allowing dynamical switching between scattering functions. Downside would be having yet another library that has to be included in the generated <instrument_name>.c - and possibly other that I haven't thought of :-)

**Comment by ebknudsen** [comment]
This thing obviously would not work for GPU. Back to the function pointer problem (fpp)

**Comment by willend** [comment]
@ebknudsen indeed. I am thinking a "code-generated" string-based switch could work, but maybe monkey-patching the sas*.c files is the most clever thing to do.

**Comment by willend** [comment]
- Or the "one component pr. model" mode

**Comment by ebknudsen** [comment]
minor point - we _could_ do a "real" integer-based switch since we actually have a SAS-models number with a definitive reference as to what it is (in terms of the sasview manual, no?
But agreed: I think the monkey-patch solution is step 1. To get something working.

**Comment by trnielsen** [comment]
I have made a prototype for McStas 3.0 which is using a switch statement for the SAS-model number. Tested and seems to be working on the DTU volta system.

After a conversation with Peter we decided to go for a dual approach:

1) Keep SasView.comp as it was in  McStas 2.x using the vector notation for the  sasmodels arguments (And adjust for McStas 3.0 changes). Here only one model nr can be include in the instrument. If the user needs to use a different model,  re-compiling is then needed.

2) Make a python support tool which will write a new SasModels.comp component having n sasmodels included.
Something like this: python make_sasmodel_comp.py models=[ 9, 10, 47, 48]

**Comment by farhi** [comment]
The current solution passing 'parX' args instead of vectors is OK.

**Comment by willend** [comment]
@jorobledo and I have a couple of good ideas / solutions that we will make prototypes for in the coming few weeks.

**Comment by willend** [comment]
Fixed via contribution of sasmodels category comps
